src/csync/csync_update.cpp

```diff
@@ -481,63 +467,56 @@ static int _csync_detect_update(CSYNC *ctx, const char *file,
       && st->instruction != CSYNC_INSTRUCTION_IGNORE
       && st->instruction != CSYNC_INSTRUCTION_UPDATE_METADATA
       && type != CSYNC_FTW_TYPE_DIR) {
-    st->child_modified = 1;
+    st->child_modified = true;
   }
-  ctx->current_fs = st;
+  ctx->current_fs = st.get();
 
-  csync_file_stat_free(tmp);
   st->inode = fs->inode;
-  st->mode  = fs->mode;
   st->size  = fs->size;
   st->modtime = fs->mtime;
   st->type  = type;
   st->etag   = NULL;
   if( fs->etag ) {
-      SAFE_FREE(st->etag);
-      st->etag  = c_strdup(fs->etag);
+      st->etag = fs->etag;
   }
-  csync_vio_set_file_id(st->file_id, fs->file_id);
+  st->file_id = fs->file_id;
   if (fs->fields & CSYNC_VIO_FILE_STAT_FIELDS_DIRECTDOWNLOADURL) {
-      SAFE_FREE(st->directDownloadUrl);
-      st->directDownloadUrl = c_strdup(fs->directDownloadUrl);
+      st->directDownloadUrl = fs->directDownloadUrl;
   }
   if (fs->fields & CSYNC_VIO_FILE_STAT_FIELDS_DIRECTDOWNLOADCOOKIES) {
-      SAFE_FREE(st->directDownloadCookies);
-      st->directDownloadCookies = c_strdup(fs->directDownloadCookies);
+      st->directDownloadCookies = fs->directDownloadCookies;
   }
   if (fs->fields & CSYNC_VIO_FILE_STAT_FIELDS_PERM) {
-      strncpy(st->remotePerm, fs->remotePerm, REMOTE_PERM_BUF_SIZE);
+      st->remotePerm = fs->remotePerm;
   }
 
   // For the remote: propagate the discovered checksum
   if (fs->checksumHeader && ctx->current == REMOTE_REPLICA) {
-      st->checksumHeader = c_strdup(fs->checksumHeader);
+      st->checksumHeader = fs->checksumHeader;
   }
 
   st->phash = h;
-  st->pathlen = len;
-  memcpy(st->path, (len ? path : ""), len + 1);
+  st->path = path;
+
+  CSYNC_LOG(CSYNC_LOG_PRIORITY_INFO, "file: %s, instruction: %s <<=", st->path.constData(),
+      csync_instruction_str(st->instruction));
 
   switch (ctx->current) {
     case LOCAL_REPLICA:
-      if (c_rbtree_insert(ctx->local.tree, (void *) st) < 0) {
-        csync_file_stat_free(st);
+      if (c_rbtree_insert(ctx->local.tree, (void *) st.release()) < 0) {
         ctx->status_code = CSYNC_STATUS_TREE_ERROR;
         return -1;
       }
       break;
     case REMOTE_REPLICA:
-      if (c_rbtree_insert(ctx->remote.tree, (void *) st) < 0) {
-        csync_file_stat_free(st);
+      if (c_rbtree_insert(ctx->remote.tree, (void *) st.release()) < 0) {
         ctx->status_code = CSYNC_STATUS_TREE_ERROR;
         return -1;
       }
       break;
     default:
       break;
   }
-  CSYNC_LOG(CSYNC_LOG_PRIORITY_INFO, "file: %s, instruction: %s <<=", st->path,
-      csync_instruction_str(st->instruction));
```

Comment:
did you just remove the instructions log?

Reply:
I moved it before the tree insert, where I release the `unique_pointer`.